File: sdk/typescript/src/reader.ts
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

import { decrypt, isEncryptedEnvelope, loadEncryptionKey } from './crypto.js';
import { CredentialNotFoundError, CredentialParseError } from './errors.js';
import type { ProviderFile, ProviderInfo } from './types.js';

const DEFAULT_CREDENTIALS_DIR = path.join(os.homedir(), '.sig', 'credentials');

let cachedKey: Buffer | null = null;

async function getEncryptionKey(): Promise<Buffer> {
    if (!cachedKey) {
        cachedKey = await loadEncryptionKey();
    }
    return cachedKey;
}

function sanitizeId(id: string): string {
    return id.replace(/[^a-zA-Z0-9._-]/g, '_');
}

export async function readProviderFile(
    providerId: string,
    credentialsDir: string = DEFAULT_CREDENTIALS_DIR,
): Promise<ProviderFile> {
    const filePath = path.join(credentialsDir, `${sanitizeId(providerId)}.json`);
    let content: string;
    try {
        content = await fs.readFile(filePath, 'utf-8');
    } catch (e: unknown) {
        if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
            throw new CredentialNotFoundError(providerId);
        }
        throw e;
    }
    try {
        let parsed: unknown = JSON.parse(content);
        if (isEncryptedEnvelope(parsed)) {
            const key = await getEncryptionKey();
            parsed = JSON.parse(decrypt(parsed, key));
        }
        const data = parsed as Record<string, unknown>;
        if (!data['providerId']) {
            throw new Error('Missing required fields');
        }
        const values = (data['values'] ?? data['credentials'] ?? {}) as Record<string, string>;
        return {
            providerId: data['providerId'] as string,
            strategy: (data['strategy'] as string) ?? '',
            updatedAt: (data['updatedAt'] as string) ?? '',
            ...(data['expiresAt'] !== undefined ? { expiresAt: data['expiresAt'] as string } : {}),
            values,
            ...(data['oauth2'] !== undefined
                ? { oauth2: data['oauth2'] as ProviderFile['oauth2'] }
                : {}),
        };
    } catch (e) {
        if (e instanceof CredentialNotFoundError) throw e;
        throw new CredentialParseError(filePath, e instanceof Error ? e : undefined);
    }
}

export async function listProviderFiles(
    credentialsDir: string = DEFAULT_CREDENTIALS_DIR,
): Promise<ProviderInfo[]> {
    let files: string[];
    try {
        files = await fs.readdir(credentialsDir);
    } catch (e: unknown) {
        if ((e as NodeJS.ErrnoException).code === 'ENOENT') return [];
        throw e;
    }
    const results: ProviderInfo[] = [];
    for (const file of files) {
        if (!file.endsWith('.json') || file.endsWith('.lock')) continue;
        try {
            const content = await fs.readFile(path.join(credentialsDir, file), 'utf-8');
            let parsed: unknown = JSON.parse(content);
            if (isEncryptedEnvelope(parsed)) {
                const key = await getEncryptionKey();
                parsed = JSON.parse(decrypt(parsed, key));
            }
            const data = parsed as Record<string, unknown>;
            if (data['providerId'] && (data['values'] || data['credentials'])) {
                results.push({
                    providerId: data['providerId'] as string,
                    strategy: (data['strategy'] as string) ?? '',
                    updatedAt: (data['updatedAt'] as string) ?? '',
                    ...(data['expiresAt'] !== undefined
                        ? { expiresAt: data['expiresAt'] as string }
                        : {}),
                });
            }
        } catch {
            /* skip unparseable files */
        }
    }
    return results;
}

File: sdk/python/src/sigcli_sdk/reader.py
import json
import re
from pathlib import Path
from typing import Optional
from .types import ProviderFile, ProviderInfo
from .errors import CredentialNotFoundError, CredentialParseError
from .crypto import is_encrypted_envelope, decrypt, load_encryption_key

DEFAULT_CREDENTIALS_DIR = Path.home() / ".sig" / "credentials"

_cached_key: Optional[bytes] = None

def _get_encryption_key() -> bytes:
    global _cached_key
    if _cached_key is None:
        _cached_key = load_encryption_key()
    return _cached_key

def _sanitize_id(provider_id: str) -> str:
    return re.sub(r"[^a-zA-Z0-9._-]", "_", provider_id)

def read_provider_file(provider_id: str, credentials_dir: Optional[Path] = None) -> ProviderFile:
    cred_dir = credentials_dir or DEFAULT_CREDENTIALS_DIR
    file_path = cred_dir / f"{_sanitize_id(provider_id)}.json"
    if not file_path.exists():
        raise CredentialNotFoundError(provider_id)
    try:
        raw_text = file_path.read_text(encoding="utf-8")
        data = json.loads(raw_text)
        if is_encrypted_envelope(data):
            key = _get_encryption_key()
            data = json.loads(decrypt(data, key))
        # Normalize: support both v2 (values) and v1 (credentials) field names
        values = data.get("values") or data.get("credentials") or {}
        return ProviderFile(
            providerId=data["providerId"],
            strategy=data.get("strategy", ""),
            updatedAt=data.get("updatedAt", ""),
            values=values,
            expiresAt=data.get("expiresAt"),
            oauth2=data.get("oauth2"),
        )
    except (json.JSONDecodeError, KeyError, TypeError, ValueError) as e:
        raise CredentialParseError(str(file_path), e) from e

def list_provider_files(credentials_dir: Optional[Path] = None) -> list[ProviderInfo]:
    cred_dir = credentials_dir or DEFAULT_CREDENTIALS_DIR
    if not cred_dir.exists():
        return []
    results: list[ProviderInfo] = []
    for fp in sorted(cred_dir.glob("*.json")):
        if fp.name.endswith(".lock"):
            continue
        try:
            raw_text = fp.read_text(encoding="utf-8")
            data = json.loads(raw_text)
            if is_encrypted_envelope(data):
                key = _get_encryption_key()
                data = json.loads(decrypt(data, key))
            if data.get("providerId") and (data.get("values") or data.get("credentials")):
                results.append(ProviderInfo(
                    providerId=data["providerId"],
                    strategy=data.get("strategy", ""),
                    updatedAt=data.get("updatedAt", ""),
                    expiresAt=data.get("expiresAt"),
                ))
        except Exception:
            continue
    return results
